### DISCUSSION TITLE
V has interoperability with C, but does it with C++?

### DISCUSSION THREAD

**Discussion by przemyslaw0** [vlang/v#18375] [Questions and Answers] [answered]
I'm asking this just for curiosity.

If has not, why not implementing this feature? Also, Objective C interoperability would be a good idea, Nim has it.

Another question, has V support for calling dynamic libraries as [Rust](https://stackoverflow.com/questions/66252029/how-to-dynamically-call-a-function-in-a-shared-library) and [Nim](https://nim-lang.org/docs/dynlib.html) have?

**Comment by medvednikov** [comment, marked as answer]
You can call C++ code just like you'd do it in C: via a C wrapper (export "C").

Dynamic libs support is coming in the near future.

**Reply by przemyslaw0** [reply]
Any plan to implement Objective C interoperability too?

**Reply by JalonSolov** [reply]
If Objective C has a C interface (like using `extern "C"` in C++), then it will work.  Otherwise it won't.

**Comment by sparky4pro** [comment]
I'm using Objective-C code with V, no problem.
Didn't try any Objective-C++ though...

**Reply by przemyslaw0** [reply]
Could you please post a little example or a snippet of how you do that?

**Reply by sparky4pro** [reply]
Here is a macOS status bar menu example:
https://github.com/vlang/v/tree/master/examples/macos_tray

Hope this helps.

**Reply by przemyslaw0** [reply]
Thanks.

Is it possible to pass Objective-C headers files (extension .mm) to V?

**Reply by sparky4pro** [reply]
As far as I know `*.mm` files are Objective-C++ files (whereas `*.m` files are Objective-C files).
Header files have the `*.h` or `*.hh` extension.

As I said earlier, I've never used/tried Objective-C++ files.

Keep in mind that V does not parse any files that are not `*.v` files (anyone can correct me if I'm wrong on this one, thanks).
This being said, all other files are passed on to the C compilers. As long they can process your additional files... you are in the clear.

I suppose your `*.mm` files will work as long as they don't have any classes in it, as V cannot use classes (C++ or Objective-C++) without C or Objective-C glue code.

**Reply by przemyslaw0** [reply]
Sorry, I thought .mm files were Objective-C, I'm planning to learn V instead of Swift/Objetive-C for iOS dev.

**Reply by sparky4pro** [reply]
If you want to write software for iOS with native widgets, you still would need to use Objective-C code to interface with Cocoa-Touch, because V cannot do it on its own.
I think Xcode's storyboards and stuff cannot be used, all the Cocoa-Touch widgets need to be coded programmatically.

You can use V for the all the other things, like for the actual application code and the glue code for Objective-C.
This is how I use V and Objective-C (Cocoa widgets coded programmatically) for macOS.

If you don't care if the widgets are native Cocoa-Touch or not, you could use VUI for the gui and compile it for iOS.
I've never used it, so I do not know how mature it is on iOS. 
Some used VUI for Android and it seemed to work...
